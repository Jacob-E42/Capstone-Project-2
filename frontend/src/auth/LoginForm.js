import React, { useCallback, useContext, useState } from "react";
import { Button, Form, FormGroup, Input, Label, FormFeedback } from "reactstrap";
import { Link, useNavigate } from "react-router-dom";
import UserContext from "../context_providers/UserContext";
import useForm from "../hooks/useForm";

const LoginForm = () => {
	console.debug("LoginForm");
	const { login } = useContext(UserContext);
	const [formData, handleChange] = useForm({
		email: "",
		password: ""
	});
	const [errors, setErrors] = useState({});
	const navigate = useNavigate();

	console.debug("LoginForm", "errors=", errors);

	const isValidEmail = useCallback(email => {
		// Perform email validation here
		// You can use regular expressions or other validation libraries
		// Return true if the email is valid, false otherwise
		return /^[\w-.]+@([\w-]+\.)+[\w-]{2,4}$/.test(email);
	}, []);

	const isValidPassword = useCallback(password => {
		// Perform password validation here
		// You can use regular expressions or other validation logic
		// Return true if the password is valid, false otherwise
		return /^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d!@#$%^&*]{8,}$/.test(password);
	}, []);

	const validateForm = useCallback(
		async data => {
			const validationErrors = {};

			if (!data.email.trim()) {
				validationErrors.email = "Email is required";
			} else if (!isValidEmail(data.email)) {
				validationErrors.email = "Invalid email address";
			}

			if (!data.password.trim()) {
				validationErrors.password = "Password is required";
			} else if (!isValidPassword(data.password)) {
				validationErrors.password =
					"Password must be at least 8 characters long and contain at least one letter and one number";
			}

			setErrors(validationErrors); // Updating the state with the constructed errors object
		},
		[setErrors, isValidEmail, isValidPassword]
	);

	const handleSubmit = useCallback(
		async e => {
			e.preventDefault();
			console.debug("handleSubmit");
			await validateForm(formData);
			if (Object.keys(errors).length === 0) {
				try {
					const result = await login(formData);
					if (result.success) {
						console.log("Form submitted:", formData);
						setErrors({});
						navigate("/profile");
					}
				} catch (err) {
					console.error(err);
				}
			}
		},
		[login, formData, errors, validateForm]
	);

	return (
		<Form onSubmit={handleSubmit}>
			<FormGroup>
				<Label for="email">Email</Label>
				<Input
					type="email"
					name="email"
					id="email"
					autoComplete="email"
					placeholder="Enter your email"
					value={formData.email}
					onChange={handleChange}
					invalid={!!errors.email}
				/>
				<FormFeedback className="inputError">{errors.email}</FormFeedback>
			</FormGroup>
			<FormGroup>
				<Label for="password">Password</Label>
				<Input
					type="password"
					name="password"
					id="password"
					autoComplete="current-password"
					placeholder="Enter your password"
					value={formData.password}
					onChange={handleChange}
					invalid={!!errors.password}
				/>
				<FormFeedback className="inputError">{errors.password}</FormFeedback>
			</FormGroup>
			<Button type="submit">Submit</Button>

			<Button
				type="button"
				tag={Link}
				to="/">
				Back
			</Button>
		</Form>
	);
};

export default LoginForm;
