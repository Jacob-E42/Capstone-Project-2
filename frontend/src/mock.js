import React, { useCallback, useState } from "react";
import UserContext from "./context_providers/UserContext";
import AlertContext from "./context_providers/AlertContext";
import Request from "./api";
import ApiRequest from "./api";
import ApiContext from "./context_providers/ApiContext";
import JournalContext from "./context_providers/JournalContext";

const demoUser = {
	id: 11,
	firstName: "testfirst",
	lastName: "testlast",
	email: "test@test.net",
	interests: ["baking", "jetskiing"]
};

const demoJournal = {
	id: 42,
	userId: 11,
	title: "Baking adventure",
	entryText:
		"Yesterday I tried baking for the first time like I've always wanted to. It was a complete disaster, but at least it was fun.",
	entryDate: "2023-07-24",
	emotions: null
};

const UserProvider = ({ children }) => {
	return <UserContext.Provider value={{ user: demoUser }}>{children}</UserContext.Provider>;
};

const AnonUserProvider = ({ children }) => {
	const [currentUser, setCurrentUser] = useState(null);

	const login = useCallback(
		async formData => {
			try {
				const request = new Request();
				const token = await request.login(formData);
				if (token) {
					setCurrentUser(demoUser);
					return { success: true };
				} else throw new Error("Log in token in missing");
			} catch (errors) {
				console.error("login failed", errors);
				return { success: false, errors };
			}
		},
		[setCurrentUser]
	);

	const signup = useCallback(
		async formData => {
			try {
				const request = new Request();
				const token = await request.signup(formData);
				console.log(token);
				if (token) setCurrentUser(demoUser);
				else throw new Error("missing sign in token");
				return { success: true };
			} catch (errors) {
				console.error("signup failed", errors);
				return { success: false, errors };
			}
		},
		[setCurrentUser]
	);

	return <UserContext.Provider value={{ currentUser, login, signup }}>{children}</UserContext.Provider>;
};

const AlertProvider = ({ children }) => {
	const [msg, setMsg] = useState("");
	const [color, setColor] = useState("primary");

	return <AlertContext.Provider value={{ msg, setMsg, color, setColor }}>{children}</AlertContext.Provider>;
};

const ApiProvider = ({ children }) => {
	const [request, setRequest] = useState(new ApiRequest());
	request.setToken(
		`eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MTEsImVtYWlsIjoidGVzdEB0ZXN0Lm5ldCIsImlhdCI6MTY5MjI5MDkxN30.8e12Cw-TMts0FmZALH61ch03kUMlW5QoCi60dlfhyWY`
	);

	return <ApiContext.Provider value={{ api: request }}>{children}</ApiContext.Provider>;
};

const JournalProvider = ({ children }) => {
	return <JournalContext.Provider value={{ journal: demoJournal }}>{children}</JournalContext.Provider>;
};

export { UserProvider, AnonUserProvider, AlertProvider, ApiProvider, JournalProvider };
