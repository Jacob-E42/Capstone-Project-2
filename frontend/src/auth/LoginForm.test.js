import React from "react";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { AlertProvider, AnonUserProvider } from "../mock";
import { useNavigate, Route, Routes } from "react-router-dom";
import { MemoryRouter } from "react-router";
import LoginForm from "./LoginForm";

jest.mock("../api", () => ({
	login: jest.fn(() => Promise.resolve("mockToken"))
}));

test("LoginForm renders without crashing", () => {
	render(
		<MemoryRouter>
			<AnonUserProvider>
				<AlertProvider>
					<LoginForm />
				</AlertProvider>
			</AnonUserProvider>
		</MemoryRouter>
	);
});

test("LoginForm matches snapshot", () => {
	const { asFragment } = render(
		<MemoryRouter>
			<AnonUserProvider>
				<AlertProvider>
					<LoginForm />
				</AlertProvider>
			</AnonUserProvider>
		</MemoryRouter>
	);
	expect(asFragment()).toMatchSnapshot();
});

test("LoginForm displays username and password inputs", () => {
	render(
		<MemoryRouter>
			<AnonUserProvider>
				<AlertProvider>
					<LoginForm />
				</AlertProvider>
			</AnonUserProvider>
		</MemoryRouter>
	);
	expect(screen.getByPlaceholderText("Enter your email")).toBeInTheDocument();
	expect(screen.getByPlaceholderText("Enter your password")).toBeInTheDocument();
});

// test("login works", async () => {
// 	const { getByLabelText, getByText } = render(
// 		<MemoryRouter initialEntries={["/login"]}>
// 			<AnonUserProvider>
// 				<LoginForm />
// 			</AnonUserProvider>
// 		</MemoryRouter>
// 	);

// 	// Fill out the form
// 	fireEvent.change(screen.getByLabelText(/email/i), { target: { value: "testuser@email.com" } });
// 	fireEvent.change(screen.getByLabelText(/password/i), { target: { value: "testpass" } });

// 	// Submit the form

// 	fireEvent.click(screen.getByText(/submit/i));

// 	// Check that you've been redirected to /profile
// 	expect(screen.getByText("Profile")).toBeInTheDocument();
// });
