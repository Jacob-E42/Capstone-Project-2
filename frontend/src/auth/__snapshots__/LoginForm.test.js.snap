// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm matches snapshot 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="email"
      >
        Email
      </label>
      <input
        autocomplete="email"
        class="form-control"
        id="email"
        name="email"
        placeholder="Enter your email"
        type="email"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="password"
      >
        Password
      </label>
      <input
        autocomplete="current-password"
        class="form-control"
        id="password"
        name="password"
        placeholder="Enter your password"
        type="password"
        value=""
      />
    </div>
    <button
      class="btn btn-secondary"
      type="submit"
    >
      Submit
    </button>
    <a
      class="btn btn-secondary"
      href="/"
      type="button"
    >
      Back
    </a>
  </form>
</DocumentFragment>
`;
